### PR TITLE
Redo nullable reference types fixes from original branch

### PR DESCRIPTION
**Pull request by jnm2**
Requested by @SimonCropp at https://github.com/shouldly/shouldly/pull/581#issuecomment-654060144

Please review a commit at a time. Most of this was cherry-picked from #581. I'm happy to find alternatives for any of the commits. There were significant problems with the structures that the three "Rewrite ..." commits address. We need to make the logic easier to follow. My commits very well may not have done this, but if not, we need to work toward something that does.

⚠ GitHub lists the commits in order of date rather than parent-child order. This means that the four cherry-picks appear together at the top instead of at the bottom of the list.

### CODE REVIEW

File: src/Shouldly/Internals/SourceCodeTextGetter.cs [hunk 1]
```diff
@@ -58,7 +58,7 @@ void ParseStackTrace(StackTrace? trace)
                 // walk up stack trace until we get to the calling method might have to be updated regularly as we find more
                 // scenarios. Alternately, it could be replaced with a more robust implementation.
                 while (currentFrame.GetMethod().DeclaringType == null ||
-                        (currentFrame.GetMethod().DeclaringType.FullName?.StartsWith("System.Dynamic") ?? false))
+                        (currentFrame.GetMethod().DeclaringType.FullName?.StartsWith("System.Dynamic", StringComparison.Ordinal) ?? false))
```
Comment (jnm2) [inline src/Shouldly/Internals/SourceCodeTextGetter.cs]: In two separate instances, StartsWith/EndsWith with the default CurrentCulture comparison in NUnit caused such bad performance for Jon Skeet that he reported it. For NUnit I'm proposing adding Fxcop Analyzers, but I'd recommend at least doing a one-time sweep.

File: src/Directory.Build.props [hunk 2]
```diff
@@ -5,4 +5,9 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (SimonCropp) [inline src/Directory.Build.props:8]: why are these required?

File: src/Directory.Build.props [hunk 3]
```diff
@@ -5,4 +5,9 @@
...
+    <PackageReference Include="TunnelVisionLabs.ReferenceAssemblyAnnotator" Version="1.0.0-alpha.154" PrivateAssets="all" />
+    <PackageDownload Include="Microsoft.NETCore.App.Ref" Version="[3.1.0]" />
+  </ItemGroup>
```
Comment (jnm2) [inline src/Directory.Build.props:11]: This brings two significant benefits:

- It reduces the number of times you have to add nullability suppressions (`!` operator) because it clues the compiler in to e.g. the fact that `string.IsNullOrEmpty` is a null check.
- It catches places where the BCL is documented to sometimes return null which would cause Shouldly to throw NullReferenceException.

File: src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs [hunk 4]
```diff
@@ -20,20 +20,23 @@ public class FirstNonShouldlyMethodFinder : ITestMethodFinder
...
+                    var callingFrame = stackTrace.GetFrame(i + Offset)
+                        ?? throw new InvalidOperationException("There is no stack frame at the specified offset from the first non-Shouldly stack frame.");
 
-            callingFrame = stackTrace.GetFrame(i + Offset - 1);
-            return new TestMethodInfo(callingFrame);
+                    return new TestMethodInfo(callingFrame);
+                }
+            }
+
+            throw new InvalidOperationException("Cannot find a non-Shouldly method in the stack trace.");
```
Comment (jnm2) [inline src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs:34]: Recommend using the split view for these sections. If any of this is too exotic, I'm more than happy to rewrite.

File: src/Directory.Build.props (hunk 2 above)
Comment (jnm2) [reply src/Directory.Build.props:8]: This brings two significant benefits:

- It reduces the number of times you have to add nullability suppressions (`!` operator) because it clues the compiler in to e.g. the fact that `string.IsNullOrEmpty` is a null check.
- It catches places where the BCL is documented to sometimes return null which would cause Shouldly to throw NullReferenceException.

Comment (SimonCropp) [reply src/Directory.Build.props:8]: ok cool. but given nuget ui doesnt yet recognize solutionlevel packages, do u think it would be better (for the short term) to make them project level to make it easier to update them?

File: src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs [hunk 5]
```diff
@@ -20,20 +20,23 @@ public class FirstNonShouldlyMethodFinder : ITestMethodFinder
...
         public TestMethodInfo GetTestMethodInfo(StackTrace stackTrace, int startAt = 0)
         {
-            var i = startAt;
-            StackFrame callingFrame;
-            do
+            for (var i = startAt; stackTrace.GetFrame(i) is { } frame; i++)
             {
-                callingFrame = stackTrace.GetFrame(i++);
-            } while (callingFrame.GetMethod().IsShouldlyMethod() || IsCompilerGenerated(callingFrame.GetMethod()));
+                if (frame.GetMethod() is { } method && !method.IsShouldlyMethod() && !IsCompilerGenerated(method))
```
Comment (SimonCropp) [inline src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs:25]: `frame.GetMethod() is { } method`  what does that syntax mean?

File: src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs (hunk 4 above)
Comment (SimonCropp) [reply src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs:34]: good point

File: src/Shouldly/Internals/SourceCodeTextGetter.cs (hunk 1 above)
Comment (SimonCropp) [reply src/Shouldly/Internals/SourceCodeTextGetter.cs]: thanks. i had no idea. is this still a problem in the current netcore?

File: src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs (hunk 5 above)
Comment (jnm2) [reply src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs:25]: It means if `frame.GetMethod()` is not null, put it in the `method` variable and evaluate true. Otherwise evaluate false.

It's a degenerate case of C# 8 property patterns. A non-degenerate case would actually have at least one property:

`if (frame.GetMethod() is { Name: "Foo" } method)`

Comment (jnm2) [reply src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs:25]: If you like, I'll rewrite as:

```cs
var method = frame.GetMethod();
if (method != null && !method.IsShouldlyMethod() && !IsCompilerGenerated(method))
```

Comment (SimonCropp) [reply src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs:25]: not at all. i am enjoying the education

File: src/Directory.Build.props (hunk 2 above)
Comment (jnm2) [reply src/Directory.Build.props:8]: I found some superfluous existing uses of `!` that can be removed now that the BCL is annotated. I'll push that soon.

Comment (jnm2) [reply src/Directory.Build.props:8]: Two commits pushed actually.

File: src/Shouldly/Internals/EqualityComparer.cs [hunk 6]
```diff
@@ -26,20 +25,11 @@ public EqualityComparer(IEqualityComparer? innerComparer = null)
...
                 return true;
 
-            // Null?
-            if (!type.IsValueType() || (type.IsGenericType() && type.GetGenericTypeDefinition().IsAssignableFrom(NullableType)))
-            {
-                if (Equals(x, null))
-                    return Equals(y, null);
-
-                if (Equals(y, null))
-                    return false;
```
Comment (jnm2) [inline src/Shouldly/Internals/EqualityComparer.cs:41]: Object.Equals returns false if *either* parameter is null. Therefore these are equivalent to `x is null` and `y is null`.

`y is null` is false if `x is null` because of the ReferenceEquals check at the beginning.

The `if (!type.IsValueType() || (type.IsGenericType() && type.GetGenericTypeDefinition().IsAssignableFrom(NullableType)))` check is a very long way to check `default(T) is null`. Neither check is needed because if `default(T)` is not null, then `x` and `y` cannot be null and therefore the null checks are no-ops. (They are even elided during jit.)

File: src/Shouldly/Internals/SourceCodeTextGetter.cs (hunk 1 above)
Comment (jnm2) [reply src/Shouldly/Internals/SourceCodeTextGetter.cs]: I'm not sure, but I do know that the semantics issue is the same and there is still perf overhead associated with doing linguistic comparisons instead of binary.

**Comment by jnm2** [comment]
@SimonCropp I'm comparing the diff between my original branch and this new branch. I see that there is a net difference that includes some missing `?`'s in the core assertions. Should I push more commits here or branch off for another PR so you can fix the master warnings sooner?

File: src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs (hunk 5 above)
Comment (jnm2) [reply src/Shouldly/Configuration/FirstNonShouldlyMethodFinder.cs:25]: Oh, great! 😀

File: src/Directory.Build.props (hunk 2 above)
Comment (jnm2) [reply src/Directory.Build.props:8]: > ok cool. but given nuget ui doesnt yet recognize solutionlevel packages, do u think it would be better (for the short term) to make them project level to make it easier to update them?

Forgot to refresh, just saw this. I am used to updating those by copying and pasting the PackageReference tag from the NuGet page, but I do get what you mean. Sounds good. I'll push a commit for this next.

(There's something slightly different than Directory.Build.props around the corner, [Directory.Packages.props](https://github.com/NuGet/Home/wiki/Centrally-managing-NuGet-package-versions), but that's not shipped yet.)

Comment (jnm2) [reply src/Directory.Build.props]: Done.

**Comment by SimonCropp** [comment]
> Should I push more commits here

lets continue here. i will hold off any changes that might conflict until you are done

File: src/Shouldly/Internals/SourceCodeTextGetter.cs (hunk 1 above)
Comment (jnm2) [reply src/Shouldly/Internals/SourceCodeTextGetter.cs]: See https://github.com/nunit/nunit/pull/3583 for Jon's comment and links elsewhere

Comment (jnm2) [reply src/Shouldly/Internals/SourceCodeTextGetter.cs]: The second one was just last month, perf hit was 5-6x

Comment (SimonCropp) [reply src/Shouldly/Internals/SourceCodeTextGetter.cs]: perhaps i should re-purpose this code https://github.com/Fody/Caseless to make `StartsWith("System.Dynamic")` compile to  `StartsWith("System.Dynamic", StringComparison.Ordinal)` :)

**Comment by jnm2** [comment]
@SimonCropp Stack_trace_is_trimmed_the_same_as_default_exception_stack_traces failed the PR validation build but it passes on my machine. Looks like master is failing, so I'll ignore for this PR if you don't think it's related.

File: src/Shouldly/Internals/SourceCodeTextGetter.cs (hunk 1 above)
Comment (jnm2) [reply src/Shouldly/Internals/SourceCodeTextGetter.cs]: That sounds cool, but there's already so much confusion around these APIs, idk about adding contextual subtleties...

**Comment by SimonCropp** [comment]
yeah there are two tests that fail intermittently. they are on my list

**Comment by SimonCropp** [comment]
and yes ignore them

**Comment by jnm2** [comment]
Alright, I'm done! It was much less than I thought at first glance.

The ShouldlyApi test doesn't start properly (I worked around it by removing the preprocessor and running the net48 build):

```
  Message: 
    System.ComponentModel.Win32Exception : The specified executable is not a valid application for this OS platform.
  Stack Trace: 
    Process.StartWithCreateProcess(ProcessStartInfo startInfo)
    Process.Start()
    Process.Start(ProcessStartInfo startInfo)
    Process.Start(String fileName, String arguments)
    DiffTool.Open(String receivedPath, String approvedPath, Boolean approvedExists) line 53
    ShouldMatchApprovedTestExtensions.ShouldMatchApproved(String actual, Func`1 customMessage, Action`1 configureOptions) line 79
    ShouldMatchApprovedTestExtensions.ShouldMatchApproved(String actual, Action`1 configureOptions) line 25
    ApprovePublicApi.ShouldlyApi() line 16
```

First guess is that UseShellExecute defaults to false on .NET Core but true on .NET Framework.

**Comment by SimonCropp** [comment]
awesome work. do you want me to squash?

**Comment by jnm2** [comment]
I appreciate you asking! Squash merges make me sad. If the goal is clean commit history, most Git tools have a first-parent view that is the same. Squashing prevents future investigators from checking out or bisecting within the merged branch a year from now. (This has happened to me and it was quite limiting.) But most important is that commits tell a story. Their granularity and messages were cultivated with some effort on my part to document themselves. In my day job and OSS work, I'm often tracing things backwards. I know what it's like to come across a squash commit. Overall, squashing feels like a loss.

However, I know other people see commits differently or deal with messier situations. No worries either way. If allowed, I'd rather squash things down to an acceptable size myself before merging. I stop force-pushing when reviewing starts so that reviewers can keep their place more easily.

Thanks for working so hard getting these PRs in!

**Comment by SimonCropp** [comment]
agreed. sometimes the commit history is valuable. sometimes it is a mess. i think this one is good enough to merge as is
